### PR TITLE
Add Delta support

### PR DESCRIPTION
**Pull request by ralphrass**
## Why? :open_book:
Add support to Delta format. Delta is an open source project that improves the concept of a data lake (or a feature store). It provides ACID transactions, time travel on data, schema evolution, upserts and a couple more features. With Delta we can improve performance of ETL jobs as well as improve management as Delta is fully supported by common tools like Databricks.

## What? :wrench:
The strategy here was to add Delta support while maintaining compatibility with older versions of ETL environments such as Databricks. However, Python version has to be updated, because recent DBR versions uses it. Still, great features of Delta are only available in versions 2.0 and above such as OPTIMIZE and VACUUM. Also, it is important to note that **only these versions of Delta support dynamic partition overwrite**, a feature of Spark that we use in incremental jobs.

The usage of Delta don't require change if the loaded data are in Parquet. The adoption of Delta is highly recommended, as it enable ACID transactions and features that transform data lakes in a modern and concise data warehouse.

- Add delta-spark as a requirement
- Add DeltaWriter as an independent writer
- Enable local tests with Delta
- Enable upserts, optimizations and vacuum

## Type of change
Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How everything was tested? :straight_ruler:
Unit testing
Databricks testing (ongoing)

## Attention Points :warning:
_Replace me for what the reviewer will need to pay attention to in the PR or just to cover any concerns after the merge._

### CODE REVIEW

File: butterfree/load/writers/delta_writer.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        - when_matched_delete_condition: it will add an operation to delete,
+            but only if this condition is true. Again, source and
+            target dataframe columns can be referred to respectively as
+            source.<column_name> and target.<column_name>
+        """
+        try:
+            full_table_name = DeltaWriter._get_full_table_name(table, database)
+
+            table_exists = client.conn.catalog.tableExists(full_table_name)
+            # table_is_delta = DeltaTable.isDeltaTable(client.conn, path)
```
Comment (mikaelsouza) [inline butterfree/load/writers/delta_writer.py:63]: Can we remove this?

File: butterfree/load/writers/delta_writer.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                provider = (
+                    pd_df.reset_index()
+                    .groupby(["col_name"])["data_type"]
+                    .aggregate("first")
+                    .Provider
+                )
+                table_is_delta = provider.lower() == "delta"
+
+                if not table_is_delta:
+                    DeltaWriter()._convert_to_delta(client, full_table_name)
```
Comment (mikaelsouza) [inline butterfree/load/writers/delta_writer.py:78]: If I understood correctly, if the table exists, we're converting it to delta, right?

Since we're using Pandas, is there any change that the Pandas representation might consume too much memory based on the size of the table or this Pandas representation is just storing some metadata that doesn't get bigger when the table gets bigger?

File: butterfree/load/writers/delta_writer.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            merge_builder.whenMatchedUpdateAll(
+                condition=when_matched_update_condition
+            ).whenNotMatchedInsertAll(
+                condition=when_not_matched_insert_condition
+            ).execute()
+        except Exception as e:
+            logger.error(f"Merge operation on {full_table_name} failed: {e}")
+
+    @staticmethod
+    def vacuum(table: str, retention_hours: int, client: SparkClient):
```
Comment (mikaelsouza) [inline butterfree/load/writers/delta_writer.py:106]: The docstring tells us the default retention time is 7 days.

Should it be set as a default value for this parameter?

File: logging.json
Comment (mikaelsouza) [inline logging.json:1]: Is this file necessary?

**Review by mikaelsouza — APPROVED** [review]

File: butterfree/load/writers/delta_writer.py (hunk 2 above)
Comment (ralphrass) [reply butterfree/load/writers/delta_writer.py:78]: Yes, we are converting. Describe table is a very simple query, it's the same resource consumption no matter the size of the table. That is a trick found to correctly identify the table type without using path.

File: logging.json
Comment (ralphrass) [reply logging.json:1]: removed

File: butterfree/load/writers/delta_writer.py (hunk 3 above)
Comment (ralphrass) [reply butterfree/load/writers/delta_writer.py:104]: This is Delta's default.
